### PR TITLE
prov/efa: fix rnr deadlock

### PR DESCRIPTION
**Pull request by wzamazon**
Fix a deadlock in RNR support

### CODE REVIEW

File: prov/efa/src/rxr/rxr_cq.c [hunk 1]
```diff
@@ -213,16 +213,17 @@ void rxr_cq_write_tx_error(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry,
 		break;
 	case RXR_TX_QUEUED_CTRL:
 	case RXR_TX_QUEUED_SHM_RMA:
-	case RXR_TX_QUEUED_REQ_RNR:
-	case RXR_TX_QUEUED_DATA_RNR:
-		dlist_remove(&tx_entry->queued_entry);
+		dlist_remove(&tx_entry->queued_ctrl_entry);
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_cq.c:219]: I don't fully understand this section, what state are the RNR packets in this switch case?

**Review by wckzhang — COMMENTED** [review]
Started reading but it doesn't seem like I'll finish any time soon, will finish tomorrow

File: prov/efa/src/rxr/rxr_cq.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c:219]: With this PR, RNR is not tracked by state, but tracked by the newly introduced flag `RXR_TX_ENTRY_QUEUED_RNR`.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c:219]: See line 224, where

```
if (tx_entry->rxr_flags & RXR_TX_ENTRY_QUEUED_RNR)
		dlist_remove(&tx_entry->queued_rnr_entry);
```

Comment (wckzhang) [reply prov/efa/src/rxr/rxr_cq.c:219]: If that's the case, wouldn't we fall into the default case and hit an assert?

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c:219]: I am not sure I understand. Where will we hit the assert?

File: prov/efa/src/rxr/rxr_ep.c [hunk 2]
```diff
@@ -642,7 +651,7 @@ static void rxr_ep_free_res(struct rxr_ep *rxr_ep)
 		tx_entry = container_of(entry, struct rxr_tx_entry,
 					queued_rnr_entry);
 		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL,
-			"Closing ep with queued tx_entry: %p\n",
+			"Closing ep with queued rnr tx_entry: %p\n",
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:657]: This should be in the first commit for tx, not here

File: prov/efa/src/rxr/rxr_ep.c [hunk 3]
```diff
@@ -1888,7 +1911,7 @@ void rxr_ep_progress_internal(struct rxr_ep *ep)
 		if (rx_entry->state == RXR_RX_FREE)
 			continue;
 
-		dlist_remove(&rx_entry->queued_entry);
+		dlist_remove(&rx_entry->queued_ctrl_entry);
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:1917]: Do you need to &= ~RXR_RX_QUEUED_CTRL here?

File: prov/efa/src/rxr/rxr_cq.c (hunk 1 above)
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_cq.c:219]: William is asking if the tx_entry is for a RNR packet, since it's not tracked by state, will it fall into the default case (L218) and hit assert(0 && "tx_entry unknown state") (L221)

File: prov/efa/src/rxr/rxr_ep.c [hunk 4]
```diff
@@ -475,6 +475,9 @@ void rxr_release_tx_entry(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry)
 	if (tx_entry->rxr_flags & RXR_TX_ENTRY_QUEUED_RNR)
 		dlist_remove(&tx_entry->queued_rnr_entry);
 
+	if (tx_entry->state == RXR_TX_QUEUED_CTRL)
```
Comment (wckzhang) [inline prov/efa/src/rxr/rxr_ep.c:478]: Okay I see you put this in a separate patch. There's no point having a bugfix patch for a bug introduced in the same PR, just squash into original commit.

**Review by wckzhang — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr_cq.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c:219]: I see. 
It will be still in one of the states RXR_TX_SEND or RXR_TX_REQ
An tx_entry in send or req state can both encounter RNR, before this patch their state will be changed to UEUE_DATA_RNR and QUEUE_CTRL_RNR. After this patch, when encounter RNR, their state remain unchanged, the flag RXR_TX_ENTRY_ENCOUTER_RNR is toggled.

So, in this code block, the state will be either RXR_TX_SEND or RXR_TX_REQ

File: prov/efa/src/rxr/rxr_ep.c [hunk 5]
```diff
@@ -472,6 +472,9 @@ void rxr_release_tx_entry(struct rxr_ep *ep, struct rxr_tx_entry *tx_entry)
 		rxr_pkt_entry_release_tx(ep, pkt_entry);
 	}
 
+	if (tx_entry->rxr_flags & RXR_TX_ENTRY_QUEUED_RNR)
+		dlist_remove(&tx_entry->queued_rnr_entry);
+
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:477]: I added a separate patch for it.

File: prov/efa/src/rxr/rxr_ep.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:1917]: There is queued_ctrl is a state, not a flag. here we change the state to RXR_RX_RECV

File: prov/efa/src/rxr/rxr_ep.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:478]: Done in new revision

File: prov/efa/src/rxr/rxr_ep.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:477]: Merged into 1st patch as you suggested in new revision.

File: prov/efa/src/rxr/rxr_cq.c [hunk 6]
```diff
@@ -117,18 +117,21 @@ void rxr_cq_write_rx_error(struct rxr_ep *ep, struct rxr_rx_entry *rx_entry,
...
-	dlist_foreach_container_safe(&rx_entry->queued_pkts,
-				     struct rxr_pkt_entry,
-				     pkt_entry, entry, tmp)
-		rxr_pkt_entry_release_tx(ep, pkt_entry);
+	if (rx_entry->rxr_flags & RXR_RX_ENTRY_QUEUED_RNR) {
+		dlist_foreach_container_safe(&rx_entry->queued_pkts,
+					     struct rxr_pkt_entry,
+					     pkt_entry, entry, tmp)
+			rxr_pkt_entry_release_tx(ep, pkt_entry);
+		dlist_remove(&rx_entry->queued_ctrl_entry);
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_cq.c]: Thanks! fixed in new revision.

File: prov/efa/src/rxr/rxr_ep.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:657]: fixed in new revision. Thank you!

**Review by wckzhang — APPROVED** [review]

File: prov/efa/src/rxr/rxr_ep.c [hunk 7]
```diff
@@ -1885,52 +1901,54 @@ void rxr_ep_progress_internal(struct rxr_ep *ep)
...
+		if (ret == -FI_EAGAIN)
+			break;
+
+		if (OFI_UNLIKELY(ret)) {
+			rxr_cq_write_tx_error(ep, tx_entry, -ret, -ret);
+			return;
 		}
+
+		dlist_remove(&tx_entry->queued_ctrl_entry);
+		if (tx_entry->state == RXR_TX_QUEUED_CTRL)
```
Comment (zhngaj) [inline prov/efa/src/rxr/rxr_ep.c:1973]: Do we really need this if condition here?

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_ep.c:1973]: We do. because `rxr_post_ctrl_once` will change `tx_entry->state` for some packet types, in the `rxr_xxx_handle_sent` function.

**Review by zhngaj — COMMENTED** [review]
In the first commit message,  I'd add something like `, which exhausts tx packet pool` at the end of situation 1. 
So one can easily figure out the failed sending of 2 is because of 1 eating up all packet pool.

There are a few typos in the 1st commit message. pls double check. 
paket->packet
will can cause 
This patch address->addresses

**Comment by wzamazon** [comment]
@zhngaj  Fixed the commit message as you suggested

**Review by zhngaj — APPROVED** [review]
